### PR TITLE
Rework new SendControlTransferAsync

### PR DESCRIPTION
**Pull request by josesimoes**
- Rework USB Device Request and related.
- Update SetupPacket accordingly.
- Add new return class ControlTransferResult.
- Update SendControlTransferAsync accordingly.
- Rework SendControlTransferAsync to better handle send and receive control transfer requests.
- Rework TestConnectToSTMDFUMode2 test.

### CODE REVIEW

File: src/Device.Net/ControlTransferResult.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+﻿namespace Device.Net
```
Comment (MelbourneDeveloper) [inline src/Device.Net/ControlTransferResult.cs:1]: @josesimoes why is this necessary?

It's exactly the same as ReadResult

File: src/Device.Net/ControlTransferResult.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        #region Public Properties
+#pragma warning disable CA1819 // Properties should not return arrays
+        public byte[] Data { get; }
+#pragma warning restore CA1819 // Properties should not return arrays
+
+        public uint BytesTransferred { get; }
+        #endregion
+
+        #region Conversion Operators
+        public byte[] ToByteArray() => Data;
```
Comment (MelbourneDeveloper) [inline src/Device.Net/ControlTransferResult.cs:16]: ReadResult has an implicit operator for converting to arrays

File: src/Usb.Net/Windows/WindowsUsbInterface.cs [hunk 3]
```diff
@@ -71,29 +71,59 @@ public void Dispose()
...
+                        // TODO
+                        transferBuffer = new byte[SetupPacket.SetupPacketSize + setupPacket.Length];
+                        // check buffer == null
+                    }
                 }
 
-                //Make a copy so we don't mess with the array passed in
-                var bufferCopy = new byte[inputBuffer.Length];
-                Array.Copy(inputBuffer, bufferCopy, inputBuffer.Length);
+                ////Make a copy so we don't mess with the array passed in
```
Comment (MelbourneDeveloper) [inline src/Usb.Net/Windows/WindowsUsbInterface.cs:101]: @josesimoes it's not a good idea to mutate the array that the user passes in.

The windows API shouldn't do that.

File: src/Usb.Net/UsbDeviceRequestType.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+
+namespace Usb.Net
+{
+    public class UsbDeviceRequestType
```
Comment (MelbourneDeveloper) [inline src/Usb.Net/UsbDeviceRequestType.cs:5]: @josesimoes this looks good.

File: src/Device.Net/ControlTransferResult.cs (hunk 1 above)
Comment (josesimoes) [reply src/Device.Net/ControlTransferResult.cs:1]: Almost... the property BytesTransferred name is different to make it clear that's the bytes transferred. 
That is properly set for both send and receive operations.

File: src/Usb.Net/Windows/WindowsUsbInterface.cs [hunk 5]
```diff
@@ -71,29 +71,59 @@ public void Dispose()
...
+                    {
+                        // data is to be received
+                        // create buffer of the appropriate size
+                        transferBuffer = new byte[setupPacket.Length];
+                    }
+                    else
+                    {
+                        // data is to be sent
+                        // TODO
+                        transferBuffer = new byte[SetupPacket.SetupPacketSize + setupPacket.Length];
```
Comment (MelbourneDeveloper) [inline src/Usb.Net/Windows/WindowsUsbInterface.cs:96]: @josesimoes why not move the array.copy code here?

File: src/Device.Net/ControlTransferResult.cs (hunk 2 above)
Comment (josesimoes) [reply src/Device.Net/ControlTransferResult.cs:16]: Yes, but the code analyser complains about that....
I'm quite all right if you want to default to that implicit operation here too.

Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:16]: We can potentially rename the class, but I'm not going to have two types that do exactly the same thing

Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:16]: @josesimoes what's the issue with analyzer?

Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:16]: It is still ReadResult whether it's from control transfer or a normal read.

File: src/Device.Net/ControlTransferResult.cs (hunk 1 above)
Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:1]: Hmm...

It's still reading the bytes from the device... Is the distinction that important?

File: src/Usb.Net/Windows/WindowsUsbInterface.cs (hunk 5 above)
Comment (josesimoes) [reply src/Usb.Net/Windows/WindowsUsbInterface.cs:96]: Still haven't tested this fully, but I think the buffer has to be done concatenating the setup packet and the data buffer.

File: src/Device.Net/ControlTransferResult.cs (hunk 1 above)
Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:1]: @josesimoes in my branch I fixed the code analyzer rule.

csharp_style_expression_bodied_operators = true:error

File: src/Usb.Net/Windows/WindowsUsbInterface.cs (hunk 5 above)
Comment (MelbourneDeveloper) [reply src/Usb.Net/Windows/WindowsUsbInterface.cs:96]: @josesimoes OK yes. We definitely need a test for this before we move forward.

File: src/Device.Net/ControlTransferResult.cs (hunk 2 above)
Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:16]: @josesimoes in my branch I fixed the code analyzer rule.

csharp_style_expression_bodied_operators = true:error

Comment (josesimoes) [reply src/Device.Net/ControlTransferResult.cs:16]: That's good!

**Comment by MelbourneDeveloper** [comment]
@josesimoes the only issue I have is ControlTransferResult, but other than that I'm happy to accept this PR. Thanks.

We do need to test sending some data as part of the control transfer so I will put the array copy part back, but I you may be right that that is not the correct way. We will need a test for that.

File: src/Device.Net/ControlTransferResult.cs (hunk 1 above)
Comment (josesimoes) [reply src/Device.Net/ControlTransferResult.cs:1]: I believe it is. Because when you are sending data it will report how many bytes where actually SENT.
And you're reading data it will tell you how many bytes where actually RECEIVED.

On both that's the gold check for the operation actually succeeding. Not only the return from the call.

Plus on the name, I feel that ReadResult is a bit... short... because it's being used for READ *AND* WRITE operations, therefore the name ControlTransferResult which works for both.

**Review by josesimoes — COMMENTED** [review]
> @josesimoes the only issue I have is ControlTransferResult, but other than that I'm happy to accept this PR. Thanks.
> 
> We do need to test sending some data as part of the control transfer so I will put the array copy part back, but I you may be right that that is not the correct way. We will need a test for that.

As soon as we are OK with this I'll start with the "sending data" part.
Just wanted to have a feedback if this was heading on the right direction before moving any further. 😉

File: src/Device.Net/ControlTransferResult.cs (hunk 1 above)
Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:1]: I could change the name to TransferResult, and the property to BytesTransferred but this would break all the existing implementations.

Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:1]: ReadResult is not being used for write operations. It's only used for reading from the device.

Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:1]: @josesimoes after reading this a few times, I'm more confused now.

When we do a control transfer, we send a buffer of data to the device. Then, the device sends the buffer back. So, technically, we may be interested in the transfer length for both operations (send and receive). However, the Windows API does not give us that. We only get the length of the data that was read back from the device. Isn't that understanding correct?

Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:1]: In all cases, what we get is a Read Result.

Comment (josesimoes) [reply src/Device.Net/ControlTransferResult.cs:1]: You're the boss... that's your call...
ReadResult is not used for write operations because there is none, yet. (I guess).
Where there is the name will be kind of awkward, no?!

Suggestion: 
1) keep the existing ReadResult and methods and add the [Obsolete] attribute to them.
2) Implement the "new" TransferResult with those calls.
3) At some future version remove the deprecated ones.

Comment (josesimoes) [reply src/Device.Net/ControlTransferResult.cs:1]: > @josesimoes after reading this a few times, I'm more confused now.
> 
> When we do a control transfer, we send a buffer of data to the device. Then, the device sends a buffer back with different data. So, technically, we may be interested in the transfer length for both operations (send and receive). However, the Windows API does not give us that. We only get the length of the data that was read back from the device. Isn't that understanding correct?

Nope... The LengthTransferred in WinUsb_ControlTransfer reports the actuall number of bytes that were SENT or RECEIVED according to the operation that was requested (RequestDirection.In or RequestDirection.Out).

Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:1]: No @josesimoes. you're pointing out something that I must be misunderstanding. I need to get this straight.

You're saying that the you want to use the structure for the result of sends, and receives, but the Windows API doesn't give us the result of sends - only receive. We only get the read result of what comes back from the device., 

This is the method signature:

```cs
public static extern bool WinUsb_ControlTransfer(IntPtr InterfaceHandle, WINUSB_SETUP_PACKET SetupPacket, byte[] Buffer, uint BufferLength, ref uint LengthTransferred, IntPtr Overlapped);
```

I could be wrong, but my understanding is that `LengthTransferred` is the length read from the device.

Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:1]: Because the operation does both a send and a receive, it's almost like we need two results.

Comment (josesimoes) [reply src/Device.Net/ControlTransferResult.cs:1]: Please check here: https://docs.microsoft.com/en-us/windows/win32/api/winusb/nf-winusb-winusb_controltransfer


Depending if you're sending data to the device or reading from it, it will tell you how many bytes were actually transferred. I insist! 😉

Comment (MelbourneDeveloper) [reply src/Device.Net/ControlTransferResult.cs:1]: > Nope... The LengthTransferred in WinUsb_ControlTransfer reports the actuall number of bytes that were SENT or RECEIVED according to the operation that was requested (RequestDirection.In or RequestDirection.Out).

OK. That's where my misunderstanding was. Thanks

File: src/Device.Net/ControlTransferResult.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
+﻿namespace Device.Net
+{
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+    public readonly struct ControlTransferResult
```
Comment (MelbourneDeveloper) [inline src/Device.Net/ControlTransferResult.cs:4]: @josesimoes let's rename ReadResult to TranserResult ,and rename BytesRead to BytesTransferred. Then we can delete this class.

You can do that in your branch, or I can change that after the PR. Let me know.

Comment (josesimoes) [reply src/Device.Net/ControlTransferResult.cs:4]: Suggest that we tackle that after this is merged, so the PR is more atomic.
